engine/widget/palette.3
.\"	$Csoft: palette.3,v 1.8 2005/01/23 11:49:13 vedge Exp $
.\"
.\" Copyright (c) 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
.\" <http://www.csoft.org>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT,
.\" INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.Dd January 30, 2003
.Dt PALETTE 3
.Os
.ds vT Agar API Reference
.ds oS Agar 1.0
.Sh NAME
.Nm palette
.Nd agar RGBA color selection widget
.Sh SYNOPSIS
.Bd -literal
#include <engine/widget/palette.h>
.Ed
.Sh DESCRIPTION
The
.Nm
widget binds to a pixel (and matching
.Xr SDL_PixelFormat 3 )
in order to manipulate its 8-bit RGBA components.
.Pp
The
.Xr hsvpal 3
widget provides an alternative to
.Nm
which is generally better suited for graphical editors.
.Sh INHERITANCE HIERARCHY
.Pp
.Xr object 3 ->
.Xr widget 3 ->
.Nm .
.Sh INITIALIZATION
.nr nS 1
.Ft "struct palette *"
.Fn palette_new "void *parent" "enum palette_type type"
.Pp
.Ft void
.Fn palette_init "struct palette *palette" "enum palette_type type"
.nr nS 0
.Pp
The
.Fn palette_new
function allocates, initializes, and attaches a
.Nm
widget.
The
.Fn palette_init
function initializes a
.Nm
widget.
.Fa type
may be one of:
.Bd -literal
enum palette_type {
	PALETTE_RGB,    /* RGB value */
	PALETTE_RGBA    /* RGBA value */
};
.Ed
.Sh BINDINGS
The
.Nm
widget provides the following bindings:
.Pp
.Bl -tag -compact -width "SDL_PixelFormat **pixel-format "
.It Va Uint32 *pixel
Pixel to read from and write to.
.It Va SDL_PixelFormat **pixel-format
Pointer to the
.Xr SDL_PixelFormat 3
to use with
.Va pixel .
.El
.Sh EVENTS
The
.Nm
widget reacts to the following events:
.Pp
.Bl -tag -compact -width 25n
.It window-mouse*
Alter the selected RGBA component.
.El
.Pp
The
.Nm
widget generates the following events:
.Pp
.Bl -tag -compact -width 2n
.It Fn palette-changed "void"
The color has been modified.
.El
.Sh SEE ALSO
.Xr agar 3 ,
.Xr widget 3 ,
.Xr window 3
.Sh HISTORY
The
.Nm
widget first appeared in Agar 1.0.


engine/widget/palette.c
/*	$Csoft: palette.c,v 1.31 2005/05/20 05:56:40 vedge Exp $	*/

/*
 * Copyright (c) 2002, 2003, 2004, 2005 CubeSoft Communications, Inc.
 * <http://www.csoft.org>
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED. IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE FOR
 * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

#include <engine/engine.h>
#include <engine/config.h>
#include <engine/view.h>

#include "palette.h"

#include <engine/widget/primitive.h>
#include <engine/widget/window.h>

const struct widget_ops palette_ops = {
	{
		NULL,		/* init */
		NULL,		/* reinit */
		widget_destroy,
		NULL,		/* load */
		NULL,		/* save */
		NULL		/* edit */
	},
	palette_draw,
	palette_scale
};

static void palette_changed(int, union evarg *);

struct palette *
palette_new(void *parent, enum palette_type type)
{
	struct palette *pal;

	pal = Malloc(sizeof(struct palette), M_OBJECT);
	palette_init(pal, type);
	object_attach(parent, pal);
	return (pal);
}

void
palette_init(struct palette *pal, enum palette_type type)
{
	int i;

	widget_init(pal, "palette", &palette_ops,
	    WIDGET_FOCUSABLE|WIDGET_WFILL|WIDGET_CLIPPING);
	widget_bind(pal, "pixel", WIDGET_UINT32, &pal->pixel);
	widget_bind(pal, "pixel-format", WIDGET_POINTER, &vfmt);

	pal->pixel = 0;
	pal->type = type; 

	switch (type) {
	case PALETTE_RGB:
		pal->nbars = 3;
		break;
	case PALETTE_RGBA:
		pal->nbars = 4;
		break;
	}

	for (i = 0; i < pal->nbars; i++) {
		pal->bars[i] = scrollbar_new(pal, SCROLLBAR_HORIZ);

		widget_set_int(pal->bars[i], "max", 255);
		event_new(pal->bars[i], "scrollbar-changed", palette_changed,
		    "%p, %i", pal, i);
	}
}

static void
palette_changed(int argc, union evarg *argv)
{
	struct palette *pal = argv[1].p;
	int nbar = argv[2].i;
	struct widget_binding *bPixel, *bFormat;
	Uint32 *pPixel;
	SDL_PixelFormat **pFormat;
	Uint8 r, g, b, a;

	bPixel = widget_get_binding(pal, "pixel", &pPixel);
	bFormat = widget_get_binding(pal, "pixel-format", &pFormat);
	SDL_GetRGBA(*pPixel, *pFormat, &r, &g, &b, &a);
	switch (nbar) {
	case 0:
		r = widget_get_int(pal->bars[nbar], "value");
		break;
	case 1:
		g = widget_get_int(pal->bars[nbar], "value");
		break;
	case 2:
		b = widget_get_int(pal->bars[nbar], "value");
		break;
	case 3:
		a = widget_get_int(pal->bars[nbar], "value");
		break;
	}
	*pPixel = SDL_MapRGBA(*pFormat, r, g, b, a);
	widget_binding_unlock(bFormat);
	widget_binding_unlock(bPixel);

	event_post(NULL, pal, "palette-changed", NULL);
}

void
palette_scale(void *p, int w, int h)
{
	struct palette *pal = p;
	int i, y = 0;

	if (w == -1 && h == -1) {
		WIDGET(pal)->w = 128;
		WIDGET(pal)->h = 0;

		for (i = 0; i < pal->nbars; i++) {
			struct widget *bar = (struct widget *)pal->bars[i];

			WIDGET_OPS(bar)->scale(bar, -1, -1);
			if (bar->w > WIDGET(pal)->w) {
				WIDGET(pal)->w = bar->w;
			}
			WIDGET(pal)->h += bar->h;

			bar->x = 0;
			bar->y = y;
			y += bar->h;
		}

		WIDGET(pal)->w += WIDGET(pal)->h;
	} else {
		int sbh = h / pal->nbars;
		int prevw = h;					/* Square */

		for (i = 0; i < pal->nbars; i++) {
			WIDGET(pal->bars[i])->x = 0;
			WIDGET(pal->bars[i])->y = i*sbh;

			widget_scale(pal->bars[i],
			    WIDGET(pal)->w - 8 - prevw,
			    sbh);
		}

		pal->rpreview.x = WIDGET(pal)->w - prevw;
		pal->rpreview.y = 0;
		pal->rpreview.w = prevw;
		pal->rpreview.h = WIDGET(pal)->h;
	}
}

void
palette_draw(void *p)
{
	char text[16];
	struct palette *pal = p;
	Uint32 c, cLabel;
	Uint8 r, g, b, a;
	SDL_Surface *label;
	struct widget_binding *bFormat;
	SDL_PixelFormat **pFormat;

	c = widget_get_uint32(pal, "pixel");
	bFormat = widget_get_binding(pal, "pixel-format", &pFormat);
	SDL_GetRGBA(c, *pFormat, &r, &g, &b, &a);
	
	widget_set_int(pal->bars[0], "value", (int)r);
	widget_set_int(pal->bars[1], "value", (int)g);
	widget_set_int(pal->bars[2], "value", (int)b);
	if (pal->nbars >= 4)
		widget_set_int(pal->bars[3], "value", (int)a);

	if (pal->type == PALETTE_RGBA)  {
		int x, y;

		primitives.tiling(pal, pal->rpreview, 16, 0,
		    COLOR(HSVPAL_TILE1_COLOR),
		    COLOR(HSVPAL_TILE2_COLOR));
		for (y = 0; y < WIDGET(pal)->h; y++) {
			for (x = 0; x < WIDGET(pal)->w; x++) {
				BLEND_RGBA2_CLIPPED(view->v,
				    WIDGET(pal)->cx+x,
				    WIDGET(pal)->cy+y,
				    r, g, b, a, ALPHA_OVERLAY);
			}
		}
	} else {
		SDL_GetRGB(c, *pFormat, &r, &g, &b);
		primitives.rect_filled(pal,
		    pal->rpreview.x, pal->rpreview.y,
		    pal->rpreview.w, pal->rpreview.h,
		    SDL_MapRGB(vfmt, r, g, b));
	}

	widget_binding_unlock(bFormat);
	
	cLabel = SDL_MapRGB(vfmt, 0, 0, 0);
	snprintf(text, sizeof(text), "%u\n%u\n%u\n%u\n", r, g, b, a);
	label = text_render(prop_get_string(config, "font-engine.default-font"),
	    10, cLabel, text);
	widget_blit(pal, label,
	    pal->rpreview.x + pal->rpreview.w/2 - label->w/2,
	    pal->rpreview.y + pal->rpreview.h/2 - label->h/2);
	SDL_FreeSurface(label);
}

